File: src/tools/coupons.ts
import { z } from "zod";
import { type ToolRegistrar, jsonResult } from "./context.js";

export const registerCouponTools: ToolRegistrar = (server, ctx) => {
  server.tool(
    "list_coupons",
    "List all coupons (paginated). Returns code, discount type/amount, " +
      "effective scope, expiration, and usage stats.",
    {},
    async () => {
      const data = await ctx.client.listCoupons();
      return jsonResult(data);
    },
  );

  server.tool(
    "get_coupon",
    "Get full details of a single coupon by its ID.",
    {
      coupon_id: z.number().int().describe("Coupon ID"),
    },
    async ({ coupon_id }) => {
      const data = await ctx.client.getCoupon(coupon_id);
      return jsonResult(data);
    },
  );

  server.tool(
    "create_coupon",
    "Create a new coupon with full configuration: code, discount type, " +
      "scope (package/category/cart), limits, date range, and more.",
    {
      code: z.string().describe("Coupon code customers will enter"),
      effective_on: z
        .enum(["package", "category", "cart"])
        .describe("What the coupon applies to"),
      packages: z
        .array(z.number().int())
        .optional()
        .describe("Package IDs the coupon applies to (when effective_on=package)"),
      categories: z
        .array(z.number().int())
        .optional()
        .describe("Category IDs the coupon applies to (when effective_on=category)"),
      discount_type: z
        .enum(["value", "percentage"])
        .describe("Whether the discount is a fixed value or a percentage"),
      discount_amount: z
        .number()
        .optional()
        .describe("Fixed discount amount (when discount_type=value)"),
      discount_percentage: z
        .number()
        .min(0)
        .max(100)
        .optional()
        .describe("Discount percentage (when discount_type=percentage)"),
      redeem_unlimited: z
        .boolean()
        .describe("If true, coupon can be redeemed unlimited times"),
      expire_never: z
        .boolean()
        .describe("If true, coupon never expires"),
      expire_limit: z
        .number()
        .int()
        .optional()
        .describe("Max number of redemptions (when redeem_unlimited=false)"),
      expire_date: z
        .string()
        .optional()
        .describe("Expiration date (when expire_never=false)"),
      start_date: z.string().optional().describe("Start date for the coupon"),
      basket_type: z
        .enum(["single", "subscription", "both"])
        .describe("Which basket types the coupon works with"),
      minimum: z
        .number()
        .min(0)
        .describe("Minimum basket value for the coupon to apply"),
      discount_application_method: z
        .number()
        .int()
        .describe("Application method: 0 = apply to each package, 1 = apply to basket total, 2 = apply once to most expensive"),
      username: z
        .string()
        .optional()
        .describe("Restrict coupon to a specific username"),
      note: z.string().optional().describe("Internal note"),
    },
    async (params) => {
      const data = await ctx.client.createCoupon(params);
      return jsonResult(data);
    },
  );

  server.tool(
    "delete_coupon",
    "Delete a coupon. Once deleted, the code can no longer be redeemed.",
    {
      coupon_id: z.number().int().describe("Coupon ID"),
    },
    async ({ coupon_id }) => {
      await ctx.client.deleteCoupon(coupon_id);
      return jsonResult({ ok: true, coupon_id, deleted: true });
    },
  );
};

File: src/tebex_mcp/tools/coupons.py
"""Coupon tools."""

from __future__ import annotations

from typing import Annotated, Any, Literal

from fastmcp import FastMCP
from pydantic import Field

from tebex_mcp.context import ToolContext
from tebex_mcp.tools._common import (
    DESTRUCTIVE,
    READ_ONLY,
    WRITE,
    map_tebex_errors,
    ok,
)


def register(mcp: FastMCP, ctx: ToolContext) -> None:
    @mcp.tool(
        name="list_coupons",
        description=(
            "List coupons (code, discount, scope, expiry, usage). Paginated: "
            "`pagination` has `currentPage`/`lastPage`/`next`; pass `page` for more."
        ),
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def list_coupons(
        page: Annotated[
            int | None, Field(description="Page number (1-indexed, default: 1)", ge=1)
        ] = None,
    ) -> Any:
        return await ctx.client.list_coupons(page)

    @mcp.tool(
        name="get_coupon",
        description="Get coupon details by ID.",
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def get_coupon(
        coupon_id: Annotated[int, Field(description="Coupon ID", ge=1)],
    ) -> Any:
        return await ctx.client.get_coupon(coupon_id)

    @mcp.tool(
        name="create_coupon",
        description="Create a coupon: code, discount type, scope, limits, date range.",
        annotations=WRITE,
    )
    @map_tebex_errors
    async def create_coupon(
        code: Annotated[str, Field(description="Code customers enter", min_length=1)],
        effective_on: Annotated[
            Literal["package", "category", "cart"],
            Field(description="What the coupon applies to"),
        ],
        discount_type: Annotated[
            Literal["value", "percentage"],
            Field(description="Fixed value or percentage"),
        ],
        redeem_unlimited: Annotated[bool, Field(description="Redeemable unlimited times")],
        expire_never: Annotated[bool, Field(description="Never expires")],
        basket_type: Annotated[
            Literal["single", "subscription", "both"],
            Field(description="Applicable basket types"),
        ],
        minimum: Annotated[float, Field(description="Min basket value to apply", ge=0)],
        discount_application_method: Annotated[
            int,
            Field(
                description="0 = each package, 1 = basket total, 2 = most expensive item",
                ge=0,
                le=2,
            ),
        ],
        packages: Annotated[
            list[int] | None,
            Field(description="Package IDs (when effective_on=package)"),
        ] = None,
        categories: Annotated[
            list[int] | None,
            Field(description="Category IDs (when effective_on=category)"),
        ] = None,
        discount_amount: Annotated[
            float | None,
            Field(description="Amount (when discount_type=value)", ge=0),
        ] = None,
        discount_percentage: Annotated[
            float | None,
            Field(description="Percent (when discount_type=percentage)", ge=0, le=100),
        ] = None,
        expire_limit: Annotated[
            int | None,
            Field(description="Max redemptions (when redeem_unlimited=false)", ge=1),
        ] = None,
        expire_date: Annotated[
            str | None, Field(description="Expiry date (when expire_never=false)")
        ] = None,
        start_date: Annotated[str | None, Field(description="Start date")] = None,
        username: Annotated[str | None, Field(description="Restrict to one username")] = None,
        note: Annotated[str | None, Field(description="Internal note")] = None,
    ) -> Any:
        return await ctx.client.create_coupon(
            {
                "code": code,
                "effective_on": effective_on,
                "packages": packages,
                "categories": categories,
                "discount_type": discount_type,
                "discount_amount": discount_amount,
                "discount_percentage": discount_percentage,
                "redeem_unlimited": redeem_unlimited,
                "expire_never": expire_never,
                "expire_limit": expire_limit,
                "expire_date": expire_date,
                "start_date": start_date,
                "basket_type": basket_type,
                "minimum": minimum,
                "discount_application_method": discount_application_method,
                "username": username,
                "note": note,
            }
        )

    @mcp.tool(
        name="delete_coupon",
        description="Delete a coupon. The code becomes unredeemable.",
        annotations=DESTRUCTIVE,
    )
    @map_tebex_errors
    async def delete_coupon(
        coupon_id: Annotated[int, Field(description="Coupon ID", ge=1)],
    ) -> dict[str, Any]:
        await ctx.client.delete_coupon(coupon_id)
        return ok(coupon_id=coupon_id, deleted=True)
